File: js/src/credentials.js
"use strict";

const fs = require("fs");
const os = require("os");
const path = require("path");
const { DEFAULT_APEX, DEFAULT_PROFILE, envFirst } = require("./defaults");

function homecloudDir() {
  const override = envFirst("HOMECLOUD_CONFIG_DIR", "HC_CONFIG_DIR");
  if (override) return path.resolve(override.replace(/^~(?=\/|\\)/, os.homedir()));
  return path.join(os.homedir(), ".homecloud");
}

function credentialsPath() {
  const override = envFirst("HOMECLOUD_CREDENTIALS_FILE", "HC_CREDENTIALS_FILE");
  if (override) return path.resolve(override.replace(/^~(?=\/|\\)/, os.homedir()));
  return path.join(homecloudDir(), "credentials");
}

function sessionPath() {
  const override = envFirst("HOMECLOUD_SESSION_FILE", "HC_SESSION_FILE");
  if (override) return path.resolve(override.replace(/^~(?=\/|\\)/, os.homedir()));
  return path.join(homecloudDir(), "session");
}

function loadCredentials(filePath = credentialsPath()) {
  if (!fs.existsSync(filePath)) {
    return { version: 2, default_profile: DEFAULT_PROFILE, profiles: {} };
  }
  const raw = JSON.parse(fs.readFileSync(filePath, "utf8"));
  if (!raw || typeof raw !== "object") throw new Error("Invalid credentials file");
  if (raw.profiles && typeof raw.profiles === "object") {
    return {
      version: Number(raw.version || 2),
      default_profile: raw.default_profile || DEFAULT_PROFILE,
      profiles: raw.profiles,
    };
  }
  // legacy flat
  return {
    version: 2,
    default_profile: DEFAULT_PROFILE,
    profiles: { [DEFAULT_PROFILE]: raw },
  };
}

function loadSession(filePath = sessionPath()) {
  if (!fs.existsSync(filePath)) return { version: 1, profiles: {} };
  const raw = JSON.parse(fs.readFileSync(filePath, "utf8"));
  return {
    version: Number(raw.version || 1),
    profiles: raw.profiles || {},
  };
}

function saveSession(session, filePath = sessionPath()) {
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(session, null, 2) + "\n", "utf8");
  try {
    fs.chmodSync(filePath, 0o600);
  } catch (_) {}
  return filePath;
}

function resolveProfile(name) {
  const creds = loadCredentials();
  const profileName =
    name ||
    envFirst("HOMECLOUD_PROFILE", "HC_PROFILE") ||
    creds.default_profile ||
    DEFAULT_PROFILE;
  const profile = creds.profiles[profileName] || {};
  const session = loadSession().profiles[profileName] || {};
  return {
    profileName,
    apex: envFirst("HOMECLOUD_APEX", "HC_APEX") || profile.apex || DEFAULT_APEX,
    accountId:
      envFirst("HOMECLOUD_ACCOUNT_ID", "HC_ACCOUNT_ID") ||
      profile.default_account_id ||
      session.active_account_id ||
      null,
    accessKeyId:
      envFirst("HOMECLOUD_ACCESS_KEY_ID", "HC_ACCESS_KEY_ID") || profile.access_key_id || null,
    secretAccessKey:
      envFirst("HOMECLOUD_SECRET_ACCESS_KEY", "HC_SECRET_ACCESS_KEY") ||
      profile.secret_access_key ||
      null,
    accessToken: session.access_token || null,
  };
}

module.exports = {
  homecloudDir,
  credentialsPath,
  sessionPath,
  loadCredentials,
  loadSession,
  saveSession,
  resolveProfile,
};

File: go/credentials.go
package homecloud

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

type profileConfig struct {
	Name             string
	Apex             string
	DefaultAccountID string
	AccessKeyID      string
	SecretAccessKey  string
}

type credentialsFile struct {
	Version        int
	DefaultProfile string
	Profiles       map[string]profileConfig
}

type profileSession struct {
	AccessToken       string
	ActiveAccountID   string
	LastUsedAccountID string
}

func homecloudDir() string {
	if override := envConfigDir(); override != "" {
		return expandHome(override)
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".homecloud")
}

func credentialsPath() string {
	if override := envCredentialsFile(); override != "" {
		return expandHome(override)
	}
	return filepath.Join(homecloudDir(), "credentials")
}

func sessionPath() string {
	if override := envSessionFile(); override != "" {
		return expandHome(override)
	}
	return filepath.Join(homecloudDir(), "session")
}

func expandHome(p string) string {
	if strings.HasPrefix(p, "~/") || strings.HasPrefix(p, "~\\") {
		home, _ := os.UserHomeDir()
		return filepath.Join(home, p[2:])
	}
	return p
}

func loadCredentialsFile(path string) (*credentialsFile, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, err
	}
	normalized := normalizeCredentials(data)
	out := &credentialsFile{
		Version:        jsonInt(normalized["version"], 2),
		DefaultProfile: jsonString(normalized["default_profile"], DefaultProfile),
		Profiles:       map[string]profileConfig{},
	}
	profiles, _ := normalized["profiles"].(map[string]any)
	for name, p := range profiles {
		pm, _ := p.(map[string]any)
		out.Profiles[name] = profileConfig{
			Name:             name,
			Apex:             jsonString(pm["apex"], platformApex()),
			DefaultAccountID: jsonString(pm["default_account_id"], ""),
			AccessKeyID:      jsonString(pm["access_key_id"], ""),
			SecretAccessKey:  jsonString(pm["secret_access_key"], ""),
		}
	}
	if len(out.Profiles) == 0 {
		return nil, newError("No profiles found in credentials file")
	}
	return out, nil
}

func normalizeCredentials(data map[string]any) map[string]any {
	if _, ok := data["profiles"]; ok {
		return data
	}
	return map[string]any{
		"version":         data["version"],
		"default_profile": DefaultProfile,
		"profiles":        map[string]any{DefaultProfile: data},
	}
}

func saveCredentialsFile(cf *credentialsFile, path string) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	profiles := map[string]any{}
	for name, p := range cf.Profiles {
		profiles[name] = map[string]any{
			"access_key_id":     p.AccessKeyID,
			"secret_access_key": p.SecretAccessKey,
		}
	}
	payload := map[string]any{
		"version":         cf.Version,
		"default_profile": cf.DefaultProfile,
		"profiles":        profiles,
	}
	raw, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	return os.WriteFile(path, raw, 0o600)
}

func loadSessionFile(path string) map[string]profileSession {
	out := map[string]profileSession{}
	raw, err := os.ReadFile(path)
	if err != nil {
		return out
	}
	var data struct {
		Profiles map[string]map[string]string `json:"profiles"`
	}
	if json.Unmarshal(raw, &data) != nil {
		return out
	}
	for name, p := range data.Profiles {
		out[name] = profileSession{
			AccessToken:       p["access_token"],
			ActiveAccountID:   p["active_account_id"],
			LastUsedAccountID: p["last_used_account_id"],
		}
	}
	return out
}

func saveSessionFile(path string, profiles map[string]profileSession) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
		return err
	}
	payloadProfiles := map[string]any{}
	for name, p := range profiles {
		if p.AccessToken == "" && p.ActiveAccountID == "" && p.LastUsedAccountID == "" {
			continue
		}
		entry := map[string]any{}
		if p.AccessToken != "" {
			entry["access_token"] = p.AccessToken
		}
		if p.ActiveAccountID != "" {
			entry["active_account_id"] = p.ActiveAccountID
		}
		if p.LastUsedAccountID != "" {
			entry["last_used_account_id"] = p.LastUsedAccountID
		}
		payloadProfiles[name] = entry
	}
	payload := map[string]any{"version": 1, "profiles": payloadProfiles}
	raw, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	raw = append(raw, '\n')
	return os.WriteFile(path, raw, 0o600)
}

func jsonString(v any, fallback string) string {
	s, _ := v.(string)
	s = strings.TrimSpace(s)
	if s == "" {
		return fallback
	}
	return strings.TrimRight(s, "/")
}

func jsonInt(v any, fallback int) int {
	switch t := v.(type) {
	case float64:
		return int(t)
	case int:
		return t
	default:
		return fallback
	}
}
